### ISSUE TITLE
Lint suggesting from_fn instead of while let Some(x)

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/rust-clippy#3940]
With the release of 1.34 with the `from_fn` function, I'd like to propose a new lint that matches: `while let Some(x) = expr` and suggests replacing it with `for x in from_fn(|| expr)`. (With a special-case for `while let Some(x) = fn()` that suggests `for x in from_fn(fn)` if possible.)

**Comment by Arnavion** [comment]
Suggesting a closure for *every* instance of `while let` will lead to lifetime problems, so this is not as straightforward as it would seem.

Eg:

```rust
fn iter(s: &mut [u8]) -> Option<(u8, &mut [u8])> {
    let (first, rest) = s.split_first_mut()?;
    Some((*first, rest))
}

fn main() {
    let mut s = &mut [1_u8, 2, 3, 4, 5][..];

    while let Some((first, rest)) = iter(s) {
        println!("{}", first);
        s = rest;
    }
}
```

Replacing it with `for (first, rest) in std::iter::from_fn(|| iter(s)) {` or `for (first, rest) in std::iter::from_fn(move || iter(s)) {` will not work.

---

I also think introducing closures where there was previously a simple `while let` is worse for readability. `std::iter::from_fn` solves the problem of creating an `Iterator` impl without the boilerplate of implementing `Iterator`, so it makes sense to use where the user would have no choice but to write an `Iterator` impl. I don't think it should be shoe-horned into `while let` loops.

**Comment by mi2ebi** [comment]
I agree with Arnavion's comment above, this doesn't really seem like a good idea to me

**Comment by jtjlehi** [comment]
@mi2ebi, do you mean it isn't a good idea for a "good first issue" or it isn't a good idea in general?
If it isn't a good idea in general, would it still possibly be worthwhile to put it in the nursery and see if it gains traction or should we just close the issue?

**Comment by mi2ebi** [comment]
In general. And I guess we *could* put it in nursery but I don't expect it to be very popular

**Comment by Jarcho** [comment]
I also don't think this is a good lint and I just want to close this. 

ping @rust-lang/clippy if there are any objections.

**Comment by Manishearth** [comment]
Yeah we'd want a stronger signal the proposed code is better and I don't think we have that.

**Comment by jtjlehi** [comment]
Thanks for the feedback!
